NOT FOR INDIVIDUAL REVIEW! Editable future graphs feature branch

chrabyrd opened the pull request:
NOT FOR INDIVIDUAL REVIEW! This branch is for the holistic review of several individual branches making up the editable_future_graphs feature

apeters commented on arches/app/views/graph.py:
```diff
@@ -538,6 +596,74 @@ def get(self, request, template="text"):
+    def post(self, request, graphid):
+        publication_id = JSONDeserializer().deserialize(request.body)
+
+        user_language = translation.get_language()
```
translation is not defined

apeters reviewed (commented):
this [line](https://github.com/archesproject/arches/pull/9613/files#diff-f435fb6934ed65137d7c2d7094c0b3e78e61ecc6aae7c6b57884d3d573f3e58eR463) is throwing an error: 
```
NameError: name 'PublishedModelError' is not defined
```

chrabyrd commented on arches/app/media/js/views/graph/graph-publication-data.js:
```diff
@@ -0,0 +1,12 @@
+        return string.replace(/,\s*}*$/, "}");
+    }
+
+    const graphPublicationDataHTML = document.querySelector('#graphPublicationData');
+    const graphPublicationData = graphPublicationDataHTML.getAttribute('graphPublicationData');
+
+    const parsedGraphPublicationData = JSON.parse(removeTrailingCommaFromObject(graphPublicationData));
+
+    return parsedGraphPublicationData;
```
agreed

chrabyrd commented on arches/app/templates/views/components/datatypes/resource-instance.htm:
```diff
@@ -24,7 +24,7 @@
     <div class="rr-table">
         <div data-bind="foreach: config.graphs" style="display: flex; flex-direction: column;">
             <div class='rr-table-row'>
-                <div class="rr-table-row-initial" data-bind="style:{color: $parent.isEditable === true || '#aaa'}">
```
out of date -- should be fixed in latest

chrabyrd commented on arches/app/models/migrations/9053_add_future_graphs.py:
```diff
@@ -0,0 +1,59 @@
+# Generated by Django 3.2.15 on 2022-11-30 16:31
```
It depends on if we're going to merge the work in branch-by-branch, or the plan is to merge in the feature branch only.

chrabyrd commented:
> this [line](https://github.com/archesproject/arches/pull/9613/files#diff-f435fb6934ed65137d7c2d7094c0b3e78e61ecc6aae7c6b57884d3d573f3e58eR463) is throwing an error:
> 
> ```
> NameError: name 'PublishedModelError' is not defined
> ```

yep, good 🎣  and fixed in latest push

apeters reviewed (changes requested):
loading a package throws and error
`no such file or directory .......templates/views/components/widgets`

also missing `...media/js/views/components/widgets`

chrabyrd commented:
> loading a package throws and error `no such file or directory .......templates/views/components/widgets`
> 
> also missing `...media/js/views/components/widgets`

Which package are you attempting to load? Could not reproduce loading the sfplanning package

apeters commented:
I copied the `pkg` folder out of the accessibility project

apeters reviewed (changes requested):
I you clone a graph and then delete that clone, the original graph become unusable and inaccessible from the UI. 
It throws a `'NoneType' object has no attribute 'get_relatable_resources'`

apeters reviewed (changes requested):
Resource type filter in search has duplicate and invalid entries.
- <img width="231" alt="Screenshot 2023-10-20 at 11 01 18 AM" src="https://github.com/archesproject/arches/assets/823773/ad6bd045-6287-4782-95dc-7b3e9285a296">

`Create new Publication` and `Revert Changes` buttons don't reflow properly on smaller screens.
- <img width="463" alt="Screenshot 2023-10-20 at 12 51 25 PM" src="https://github.com/archesproject/arches/assets/823773/b2fb1cb8-3b05-489a-9c3a-ac9daa19eff0">

Creating a new model creates 2 entries in the model hist page.  That doesn't seem right.
The model below has only been published once, but there are 2 entries.
- <img width="687" alt="Screenshot 2023-10-20 at 1 04 09 PM" src="https://github.com/archesproject/arches/assets/823773/3a3f72c7-abaf-4cf4-8139-36fb19470f10">

Reports appear to be broken.  They don't show any data.

apeters reviewed (changes requested):
if you make an edit to the current graph that's active, save it, and then revert the change, the graph changes to the `inactive` state.  See [this](https://recordit.co/A0dxssiGXC).

apeters reviewed (changes requested):
Adding a new language and then updating the a widget label on the current graph (not publishing a new graph) throws an error when viewing the report or editor of an instance of that graph.

apeters reviewed (commented):
I don't know if this is a bug or a feature, although it feels like a bug.  I noticed that when editing the current graph once you've saved change but not applied the change that the endpoint to edit a resource instance redirects (403) to the report.  From the end users perspective it feels like something is broken.

chrabyrd commented:
> I you clone a graph and then delete that clone, the original graph become unusable and inaccessible from the UI. It throws a `'NoneType' object has no attribute 'get_relatable_resources'`

good catch, ended up being an insidious little bug -- the publication is not erased on graph copy

chrabyrd commented:
> Resource type filter in search has duplicate and invalid entries.
> 
> * <img alt="Screenshot 2023-10-20 at 11 01 18 AM" width="231" src="https://user-images.githubusercontent.com/823773/277032193-ad6bd045-6287-4782-95dc-7b3e9285a296.png">

fixed in latest

> `Create new Publication` and `Revert Changes` buttons don't reflow properly on smaller screens.
> 
> * <img alt="Screenshot 2023-10-20 at 12 51 25 PM" width="463" src="https://user-images.githubusercontent.com/823773/277039355-b2fb1cb8-3b05-489a-9c3a-ac9daa19eff0.png">

Yeah this fell out of a conversation w/Rob Dennis and I. The UX will definitely need further tweaking but I say punt until operational && we can have another UX discussion.

> Creating a new model creates 2 entries in the model hist page. That doesn't seem right. The model below has only been published once, but there are 2 entries.
> 
> * <img alt="Screenshot 2023-10-20 at 1 04 09 PM" width="687" src="https://user-images.githubusercontent.com/823773/277041551-3a3f72c7-abaf-4cf4-8139-36fb19470f10.png">

Ah this is by design. As of this branch a graph _cannot_ exist without a publication. So when a graph is first created saved, it gets a publication.

> Reports appear to be broken. They don't show any data.

fixed in latest

chrabyrd commented:
> this

We should have a chat about this, there's some complexity here.

chrabyrd commented:
> Adding a new language and then updating the a widget label on the current graph (not publishing a new graph) throws an error when viewing the report or editor of an instance of that graph in the new language.

Can you expand on this, maybe include a video?

chrabyrd commented:
> I don't know if this is a bug or a feature, although it feels like a bug. I noticed that when editing the current graph once you've saved change but not applied the change that the endpoint to edit a resource instance redirects (403) to the report. From the end users perspective it feels like something is broken.

This is a feature, happy to chat about i.

apeters commented:
Exporting a card as a branch throws an error. `'dict' object has no attribute 'is_immutable'`

apeters commented:
Branches that are "Active" still show up as "inactive" in the append branch window.![Screenshot 2023-12-08 at 3 52 07 PM](https://github.com/archesproject/arches/assets/823773/59045a06-549b-4ee7-8886-a245b50ba47a)

apeters commented:
It would be nice to see the counts of resources that are associated with each published graph.  It's hard to know which graph has data associated with it.
![Screenshot 2023-12-08 at 4 20 01 PM](https://github.com/archesproject/arches/assets/823773/3c1d751d-6695-47eb-98a8-57af568a3980)

apeters commented:
a couple of graph properties don't get persisted when re-publishing an existing graph:
`resource models that may be related` and `slug`
![Screenshot 2023-12-08 at 5 55 49 PM](https://github.com/archesproject/arches/assets/823773/70044106-9620-4318-a772-8df759b324fb)
![Screenshot 2023-12-08 at 5 55 43 PM](https://github.com/archesproject/arches/assets/823773/a1c427cd-8f52-4ff5-a1c3-377baf509627)

chrabyrd commented:
> Exporting a card as a branch throws an error. `'dict' object has no attribute 'is_immutable'`

fixed in latest

chrabyrd commented:
> Branches that are "Active" still show up as "inactive" in the append branch window.![Screenshot 2023-12-08 at 3 52 07 PM](https://private-user-images.githubusercontent.com/823773/289212513-59045a06-549b-4ee7-8886-a245b50ba47a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI0MjY0NjQsIm5iZiI6MTcwMjQyNjE2NCwicGF0aCI6Ii84MjM3NzMvMjg5MjEyNTEzLTU5MDQ1YTA2LTU0OWItNGVlNy04ODg2LWEyNDViNTBiYTQ3YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxM1QwMDA5MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNjhmYjA0MDhhMGMxNDc2NDFlM2RmMzhjMjA0ZWE1YTQzMmQ5Y2QwYWVkMzU0NjRkNDRkMDVhMDBkNjZhY2UwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.TgbuvCohLEsUEEzp06UAZ52z9VgRF6M0NjXzc-KgZQU)

fixed in latest

chrabyrd commented:
> It would be nice to see the counts of resources that are associated with each published graph. It's hard to know which graph has data associated with it. ![Screenshot 2023-12-08 at 4 20 01 PM](https://private-user-images.githubusercontent.com/823773/289216978-3c1d751d-6695-47eb-98a8-57af568a3980.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI0MjY0NjQsIm5iZiI6MTcwMjQyNjE2NCwicGF0aCI6Ii84MjM3NzMvMjg5MjE2OTc4LTNjMWQ3NTFkLTY2OTUtNDdlYi05OGE4LTU3YWY1NjhhMzk4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxM1QwMDA5MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lODQ3OGY0NDhmNzNmMmIwOTA3YjM2NTkyYzkyMGE2ZDI4ZTM2N2U1Mzc3OTVjOThjZjdmMWYyNzk1NmZmYjlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.pA9mzoAXbhH2uTMM49hldPVZsxIJLMkZvO6-eq7Jokk)

added in latest

apeters reviewed (commented):
I found an issue that, unfortunately, I can't reproduce but has happened twice on separate occasions.   I believe I was editing the current graph and only focusing on the graph name itself, testing with different languages, saving edits, undoing edits, etc..., just trying to find any issues when this issue came up.

Here is the detail of the error and additional context:
```
AttributeError
'NoneType' object has no attribute 'get_relatable_resources'
/Users/alexei/Work/Projects/arches/env_8x/arches/arches/app/models/graph.py, line 1596, in serialize
arches.app.views.graph.GraphDesignerView
/Users/alexei/Work/Projects/arches/env_8x/bin/python3
```

<img width="1021" alt="Screenshot 2024-02-08 at 4 40 31 PM" src="https://github.com/archesproject/arches/assets/823773/bd1258e4-a167-40e5-8f1e-47dc98855acc">

jacobtylerwalls commented on arches/app/views/graph.py:
```diff
@@ -210,23 +225,15 @@ def get(self, request, graphid):
         if self.graph.ontology is not None:
```
Clicking on an existing graph in AFS throws on this line because self.graph is None (in turn, because self.editable_future_graph is None, earlier). I ran the `manage.py graph publish --update` command.

chrabyrd replied:
This is a good catch. `graph publish --update` is really only useful in germs of adding or updating languages. There wasn't a great method to create editable_future_graphs in bulk for a project upgrade. So I added the management command `create_editable_future_graphs` to accommodate

jacobtylerwalls commented on arches/app/views/graph.py:
```diff
@@ -449,19 +468,13 @@ def post(self, request, graphid=None):
         if self.action == "delete_node":
             data = JSONDeserializer().deserialize(request.body)
             try:
                 graph = Graph.objects.get(graphid=graphid)
```
Clicking the delete node button just after publishing a graph (before dismissing the success banner) raises a 500. This seems unlikely, but I could see it happening if a user made a mistake and didn't bother dismissing the banner.


https://github.com/archesproject/arches/assets/38668450/84226bee-dd78-4379-af4b-3f7601f4b1cc

jacobtylerwalls reviewed (commented):
Hi! Just pulled this down to test drive it. I'm not planning on combing the diff here, but I did glance at models.py just to get a flavor of how my database was going to change on migrating. In doing that, I noticed one thing that I've tangled with before, and wanted to share a solution.

jacobtylerwalls commented:
In this workflow, nodes are created without an alias:

1. Graph Designer > New Resource Model. Edit name of top node.
2. Save. At this point the "create publication" and "revert changes" buttons appear.
3. Add a child node. Alias placeholder text appears.
4. Save Edits. Node alias is generated.
5. Add a grandchild node. Select it in sidebar. Alias placeholder text appears.
6. Discard Edits.
7. Create new publication.

Result: grandchild node has null node alias

chrabyrd commented:
> I found an issue that, unfortunately, I can't reproduce but has happened twice on separate occasions. I believe I was editing the current graph and only focusing on the graph name itself, testing with different languages, saving edits, undoing edits, etc..., just trying to find any issues when this issue came up.
> 
> Here is the detail of the error and additional context:
> 
> ```
> AttributeError
> 'NoneType' object has no attribute 'get_relatable_resources'
> /Users/alexei/Work/Projects/arches/env_8x/arches/arches/app/models/graph.py, line 1596, in serialize
> arches.app.views.graph.GraphDesignerView
> /Users/alexei/Work/Projects/arches/env_8x/bin/python3
> ```
> 
> <img alt="Screenshot 2024-02-08 at 4 40 31 PM" width="1021" src="https://private-user-images.githubusercontent.com/823773/303521402-bd1258e4-a167-40e5-8f1e-47dc98855acc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzODU5ODAsIm5iZiI6MTcxMTM4NTY4MCwicGF0aCI6Ii84MjM3NzMvMzAzNTIxNDAyLWJkMTI1OGU0LWExNjctNDBlNS04ZjFlLTQ3ZGM5ODg1NWFjYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNVQxNjU0NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05YWYyMzk3NTQwMDVkNGNkOThjYzBlY2MzNmEwNmE1OGY3OGNlNmYyNWI3ZmU3NTBjMGRjMjRlZmI5YjVjNzg2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.8MzAOWT4oriVabDr4B_mDdpXXnQDfd-U7ZN-Y0Cshlg">

I can shed a _little_ more light into this as it came up when I was chasing other issues. That happens when something has caused the graph to not properly update/serialize previously, but making it far enough in the process to reassign the `root` node. I've made some adjustments to `update_from_editable_future_graph` that have allowed the `transaction.atomic()` block to be re-added. This should hopefully correct any instances of this behavior

chrabyrd commented on arches/app/etl_modules/import_single_csv.py:
```diff
@@ -36,7 +36,7 @@ def get_graphs(self, request):
             GraphModel.objects.all()
             .exclude(pk=settings.SYSTEM_SETTINGS_RESOURCE_MODEL_ID)
             .exclude(isresource=False)
-            .exclude(publication_id__isnull=True)
```
Good catch. With this feature a Graph is _always_ published unless it's an editable future graph. I've updated all filter calls throughout the app to instead key on the `source_identifier__isnull` bool

chrabyrd commented on arches/app/models/models.py:
```diff
@@ -106,6 +102,11 @@ def __init__(self, *args, **kwargs):
             self.cardid = uuid.UUID(self.cardid)
 
+    def save(self, *args, **kwargs):
+        if self.pk == self.source_identifier_id:
+            self.source_identifier_id = None
```
++

chrabyrd commented:
> In this workflow, nodes are created without an alias:
> 
> 1. Graph Designer > New Resource Model. Edit name of top node.
> 2. Save. At this point the "create publication" and "revert changes" buttons appear.
> 3. Add a child node. Alias placeholder text appears.
> 4. Save Edits. Node alias is generated.
> 5. Add a grandchild node. Select it in sidebar. Alias placeholder text appears.
> 6. Discard Edits.
> 7. Create new publication.
> 
> Result: grandchild node has null node alias

This appears to be an issue outside of the editable_future_graph work. It looks like a node alias is not generated until the `Save Edits` button is pushed. I'll spin up a separate ticket but IMO this is out of scope

jacobtylerwalls commented on arches/app/models/models.py:
```diff
@@ -1138,7 +1184,7 @@ class TileModel(models.Model):  # Tile
     parenttile = models.ForeignKey("self", db_column="parenttileid", blank=True, null=True, on_delete=models.CASCADE)
     data = JSONField(blank=True, null=True, db_column="tiledata")
-    nodegroup = models.ForeignKey(NodeGroup, db_column="nodegroupid", on_delete=models.CASCADE)
+    nodegroup_id = models.UUIDField(db_column="nodegroupid", null=True)
```
Heads up that the first failing test on the last push looks like it was due to this change, so let's add a breaking changes notice. There could be applications that need to check their views & tests.

jacobtylerwalls commented on tests/base_test.py:
```diff
@@ -118,7 +119,11 @@ def setUpClass(cls):
             jwt_algorithm=test_settings.JWT_ALGORITHM,
         )
-        cursor.execute(sql)
+
+        try:
+            cursor.execute(sql)
+        except psycopg2.errors.UniqueViolation:  # entry already in database
+            pass
```
Hopefully this won't be a problem after #10711 but I can recheck it

jacobtylerwalls replied:
I looked into this a little further. This is still a problem after #10711. It's related to the TODO statements I added showing where we have some package loads in test case setup that _depend_ on running _outside_ of a transaction, which isn't ideal:

https://github.com/archesproject/arches/blob/350e4a7cd2cca3b17176cf31bbcbf5b5282ff695/tests/views/api_tests.py#L41-L44

I'll open an issue and ref them in the TODO statements. If we fix that, I expect we wouldn't need this workaround.

However, I don't think this workaround is doing what it appears. When I revert the bugfix that causes the api test to fail, and I run `python3 manage.py test tests.views.api_tests tests.views.auth_tests --settings=tests.test_settings`, this except isn't reached because Django reraises `UniqueViolation` as `IntegrityError`. However, if I change this to catch IntegrityError, then it causes far more failures because the transaction is aborted and doesn't clean up after itself and all subsequent tests fail.

I'm feeling like we should just remove the workaround and wait for the follow-up issue to be fixed. WDYT?

chrabyrd replied:
Hmm let's keep this fix for now but you should request changes on this PR so it can't get merged in. That way the fix is still there in case the followup doesn't work or @apeters decides to merge it in knowing that the followup is not far behind.

jacobtylerwalls reviewed (changes requested):
Marking changes requested to ensure we discuss setUpClass workaround before merge

jacobtylerwalls commented on arches/management/commands/graph.py:
```diff
@@ -73,34 +74,57 @@ def add_arguments(self, parser):
+            print("\nCreating editable_future_graph for %s" % graph.name)
+            graph.create_editable_future_graph()
+
+            print("\n%s has been updated. Creating a new publication for %s" % graph.name, graph.name)
```
Ran this command and got:
```py
  File "/web_root/arches/arches/management/commands/graph.py", line 101, in create_editable_future_graphs
    print("\n%s has been updated. Creating a new publication for %s" % graph.name, graph.name)
TypeError: not enough arguments for format string
```

jacobtylerwalls commented:
Ran the AfS view tests against this branch and found some `TileModel.nodegroup` references to fix. That gave me the idea to grep for things in core arches possibly not exercised by a test.

This looks relevant. It's fine if nodegroup returns something, but if it's None, then `.pk` will obviously raise.
https://github.com/archesproject/arches/blob/a1c2b429f5aaa54af95a4111862eb87d56709107/arches/app/utils/data_management/resources/formats/rdffile.py#L828

Also, we should update this to use nodegroup_id so we don't reintroduce the db queries that #10546 optimized out. It's not a pure replace; I figure we'll have to extract the pk's from the set of model objects.

https://github.com/archesproject/arches/blob/d4d960801140e3f62a40869c55910de6c7316ab2/arches/app/models/resource.py#L260

Resource.copy() is currently broken (#10712), but let's fix the breaking change here while we're fresh:

https://github.com/archesproject/arches/blob/a1c2b429f5aaa54af95a4111862eb87d56709107/arches/app/models/resource.py#L736

This is a little paranoid, but we could add some safety against null nodegroups here. Speaking of, are you allowed to create future versions of the system settings graph?

https://github.com/archesproject/arches/blob/a1c2b429f5aaa54af95a4111862eb87d56709107/arches/app/models/system_settings.py#L119-L120
https://github.com/archesproject/arches/blob/a1c2b429f5aaa54af95a4111862eb87d56709107/arches/app/models/system_settings.py#L130-L133

Another null guard:
https://github.com/archesproject/arches/blob/a1c2b429f5aaa54af95a4111862eb87d56709107/arches/app/utils/data_management/resources/formats/csvfile.py#L1101

Another null guard:
https://github.com/archesproject/arches/blob/a1c2b429f5aaa54af95a4111862eb87d56709107/arches/app/utils/data_management/resources/formats/rdffile.py#L206-L230

jacobtylerwalls commented on tests/models/graph_tests.py:
```diff
@@ -1071,24 +1072,130 @@ def test_appending_a_branch_with_an_invalid_ontology_class(self):
         with self.assertRaises(GraphValidationError) as cm:
             graph.save()
 
-    def test_appending_to_published_graph(self):
-        graph = Graph.objects.get(node=self.rootNode)
-        admin = User.objects.get(username="admin")
-        graph.publish(user=admin)
-        self.addCleanup(graph.unpublish)
-
-        with self.assertRaises(GraphValidationError) as cm:
-            graph.append_node()
-        self.assertEqual(cm.exception.code, 1012)
-        with self.assertRaises(GraphValidationError) as cm:
-            graph.append_branch("http://www.nasa.gov/", graphid=self.NODE_NODETYPE_GRAPHID)
-        self.assertEqual(cm.exception.code, 1012)
-
-    def test_appending_published_branch_to_unpublished_graph(self):
-        graph = Graph.objects.get(node=self.rootNode)
-        admin = User.objects.get(username="admin")
-        branch = Graph.objects.get(graphid=self.NODE_NODETYPE_GRAPHID)
-        branch.publish(user=admin)
-        self.addCleanup(branch.unpublish)
-
-        graph.append_branch("http://www.nasa.gov/", graphid=self.NODE_NODETYPE_GRAPHID)
```
Are these tests being deleted because the sense of what they are doing is captured in the tests that are being added?

chrabyrd replied:
Nah they're being deleted for relevance. It's not longer possible to append a node to a publish graph, and there's no such thing as an unpublished graph after this feature

jacobtylerwalls commented:
[stream of consciousness review, sorry...]

Re: the part of the comment above about `.nodegroup` adding queries now that's it's a property -- it occurs to me that now we won't be able to `select_related("nodegroup")`'able to avoid N+1 queries. Looping over tiles and accessing `.nodegroup` will hit the DB every time. (That is, without some ORM gymnastics like adding a custom model manager or doing a highly custom prefetch.)

If all we need is the nullability, have we considered just using making the existing foreign key nullable? Then the following would still work without breaking changes, other than needing to check for None:
 - `select_related("nodegroup")`
 - `.nodegroup = ...`
 - `.filter(nodegroup= ... )`

jacobtylerwalls commented on releases/8.0.0.md:
```diff
@@ -0,0 +1,37 @@
+
+### Upgrading an Arches project
+
+1. Create editable_future_graphs for your Resource Models using the command `python manage.py graph create_editable_future_graphs`. This will publish new versions of each Graph.
```
For a graph with a slug, after running this command, I cannot reverse migrate:

```py
django.db.utils.IntegrityError: could not create unique index "graphs_slug_32b94acf_uniq"
DETAIL:  Key (slug)=(my_slug) is duplicated.
```

chrabyrd replied:
Could you include the steps to repro please?

jacobtylerwalls replied:
1. Graph Manager > Add a slug and save (field is called "URI Slug for API Access")
2. Create new publication
3. Check db shell: you should have 2 graph rows with the slug you created
4. Reverse migrate: `python manage.py migrate models 9511`

Result: django attempts to reenforce the unique constraint on slug, but PG complains that multiple rows exist

I found this in a more roundabout way just loading AfS on 7.6 and upgrading to 8.0 and running `manage.py graph create_editable_future_graphs` which gave me the duplicate slugs.

chrabyrd replied:
This was a good catch && took some thinking on how to work around it. Should be fixed in latest

jacobtylerwalls commented on arches/app/templates/views/graph/model-history.htm:
```diff
@@ -0,0 +1,195 @@
+                    </div>
+                    <div 
+                        style="
+                            overflow-y: scroll;
```
```suggestion
                            overflow-y: auto;
```
Seems to prevent this situation on MacOS:
<img width="782" alt="always-scrollbar" src="https://github.com/archesproject/arches/assets/38668450/2a936b98-8488-4932-86bb-3cc51a5af01c">

chrabyrd replied:
can't seem to repro but setting it to auto doesn't affect the UX negatively for me, so happy to include this

jacobtylerwalls commented:
1. Create new graph
2. Click "Make changes without publishing"
3. make edit; click "Save edits"
4. Quit Designer
5. Graph Manager > Delete the original graph from "Manage..." dropdown
6. Manager refreshes to show only the graph marked "(Has unpublished changes)" > Delete that one from "Manage..." dropdown

Raises `Graph.DoesNotExist`

jacobtylerwalls commented on arches/app/views/graph.py:
```diff
@@ -210,23 +225,15 @@ def get(self, request, graphid):
         if self.graph.ontology is not None:
             branch_graphs = branch_graphs.filter(ontology=self.graph.ontology)
```
1. Create a branch from scratch (in graph designer)
2. Save Edits
3. Quit Designer
4. [You're on Branches tab of Graph Designer:] Click on that branch


Raises

```py
  File "/Users/jwalls/prj/arches/arches/app/views/graph.py", line 225, in get
    if self.graph.ontology is not None:
       ^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'ontology'
```

chrabyrd replied:
I also cannot seem to repro this either. Happy to chat if you wanna show off your workflow && compare

chrabyrd commented:
> [stream of consciousness review, sorry...]
> 
> Re: the part of the comment above about `.nodegroup` adding queries now that's it's a property -- it occurs to me that now we won't be able to `select_related("nodegroup")` to avoid N+1 queries. Looping over tiles and accessing `.nodegroup` will hit the DB every time. (That is, without some ORM gymnastics like adding a custom model manager or doing a highly custom prefetch.)
> 
> If all we need is the nullability, have we considered just making the existing foreign key nullable? Then the following would still work without breaking changes, other than needing to check for None:
> 
> * `select_related("nodegroup")`
> * `.nodegroup = ...`
> * `.filter(nodegroup= ... )`

No worries! Always glad to have to re-think/defend decisions so the best way forward is implemented. Ended up fiddling around with the code a bit to remind myself why the choice was made.

The changes made here are to support the switching of Graphs between versions. eg:

1. Create a Graph with a single string node, named 'AAA'. It's a de-facto nodegroup.
2. Create a Resource from the Graph, it has a Tile referencing 'AAA". 
3. Edit the Graph and delete the 'AAA' node. Then add a 'BBB' number node. 
4. Attempt to publish the new Graph.

An error will occur if the link between the Nodegroup/Tile still exists because in Postgres you can't have a ForiegnKey reference a non-existent entity. 

So let's say we make `Tile.nodegroup` nullable with the `on_delete=models.SET_NULL` arg. Going through the above workflow will probably work, but the only linkage between NodeGroup and Tile is  on the tile. Once the new graph is published, the 'AAA' nodegroup will be deleted and the Tile will effectively be orphaned. Attempting to view the resource report or edit the resource in the editor will end in error. 

A different solution I imagined was having both nodegroup and nodegroup_id fields on the TileModel. We could use Django signals or something so that whenever a NodeGroup is created the Tile table could be iterated and that link re-established. However it's easy for the Tiles table to grow large and I think that would be a noticeable performance hit.

So I landed on the current solution. Probably not ideal but the best path I can see right now. As far as the nullguarding goes I'm inclined to just built an Exception into the `Tile.nodegroup` property instead of playing whack-a-mole with the callsites. WDYT?

chrabyrd commented:
> 1. Create new graph
> 2. Click "Make changes without publishing"
> 3. make edit; click "Save edits"
> 4. Quit Designer
> 5. Graph Manager > Delete the original graph from "Manage..." dropdown
> 6. Manager refreshes to show only the graph marked "(Has unpublished changes)" > Delete that one from "Manage..." dropdown
> 
> Raises `Graph.DoesNotExist`

I cannot seem to repro. Are there any specific changes you're making?

![1hSUFL2hFc](https://github.com/archesproject/arches/assets/9162695/b6b1271d-59af-4e3a-987d-a449c70dc3ce)

jacobtylerwalls commented:
> Are there any specific changes you're making?

Now I can't reproduce either. No worries.

jacobtylerwalls commented on arches/app/views/api.py:
```diff
@@ -1095,8 +1146,28 @@ def get(self, request, resourceid):
 
         resource = Resource.objects.get(pk=resourceid)
-        graph = Graph.objects.get(graphid=resource.graph_id)
+        published_graph = models.PublishedGraph.objects.get(publication=resource.graph_publication, language=translation.get_language())
```
Found that an import went missing here (assuming after tons of merges). Reproduced by clicking "View Report" button from Edit Log activity:

```py
2024-04-02 16:52:28   File "/web_root/arches/arches/app/views/api.py", line 1149, in get
2024-04-02 16:52:28     published_graph = models.PublishedGraph.objects.get(publication=resource.graph_publication, language=translation.get_language())
2024-04-02 16:52:28 NameError: name 'translation' is not defined
```

jacobtylerwalls commented:
Thanks for the walkthrough on the missing nodegroup flow, I get it now.

> As far as the nullguarding goes I'm inclined to just built an Exception into the Tile.nodegroup property instead of playing whack-a-mole with the callsites. WDYT?

I think I was wrongly assuming that some of these other workflows should still succeed if there are unmigrated historical tiles pointing to nonexistent nodes. But now that I see we want them to fail, then, yeah, let's not play whackamole. As you suggested something really simple like this would do:

```diff
diff --git a/arches/app/models/models.py b/arches/app/models/models.py
index aa5a0d6e3..909e9ce5a 100644
--- a/arches/app/models/models.py
+++ b/arches/app/models/models.py
@@ -1199,8 +1199,7 @@ class TileModel(models.Model):  # Tile
 
     @property
     def nodegroup(self):
-        nodegroup_query = NodeGroup.objects.filter(pk=self.nodegroup_id)
-        return nodegroup_query.first()
+        return NodeGroup.objects.get(pk=self.nodegroup_id)
 
     def is_fully_provisional(self):
         return bool(self.provisionaledits and not any(self.data.values()))
```

jacobtylerwalls commented:
@robgaston raised the idea of making a small effort to handle other locations (e.g. search export) that naively assume a tile's nodegroup always exists -- and wrap with something more appropriate than a bare error like the below. ("Error: ad-hoc graph migration; contact your administrator" or whatever). But I can also see that being follow-up work.

I'm not totally certain what the UX is like when Debug=False for 500s like this.

```py
Traceback (most recent call last):
  File "/web_root/arches/arches/app/search/search_export.py", line 264, in get_node
    return self.node_lookup[nodeid]

During handling of the above exception ('52aa2007-c450-11e9-b5d4-a4d18cec433a'), another exception occurred:
  File "/usr/local/lib/python3.10/dist-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.10/dist-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/django/contrib/auth/decorators.py", line 23, in _wrapper_view
    return view_func(request, *args, **kwargs)
  File "/web_root/arches/arches/app/views/search.py", line 278, in export_results
    export_files, export_info = exporter.export(format, report_link)
  File "/web_root/arches/arches/app/search/search_export.py", line 160, in export
    resource_obj = self.flatten_tiles(
  File "/web_root/arches/arches/app/search/search_export.py", line 313, in flatten_tiles
    node = self.get_node(nodeid)
  File "/web_root/arches/arches/app/search/search_export.py", line 266, in get_node
    self.node_lookup[nodeid] = models.Node.objects.get(pk=nodeid)
  File "/usr/local/lib/python3.10/dist-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/django/db/models/query.py", line 637, in get
    raise self.model.DoesNotExist(

Exception Type: DoesNotExist at /search/export_results
Exception Value: Node matching query does not exist.
```

chrabyrd commented:
> @robgaston raised the idea of making a small effort to handle other locations (e.g. search export) that naively assume a tile's nodegroup always exists -- and wrap with something more appropriate than a bare error like the below. ("Error: ad-hoc graph migration; contact your administrator" or whatever). But I can also see that being follow-up work.
> 
> I'm not totally certain what the UX is like when Debug=False for 500s like this.
> 
> ```python
> Traceback (most recent call last):
>   File "/web_root/arches/arches/app/search/search_export.py", line 264, in get_node
>     return self.node_lookup[nodeid]
> 
> During handling of the above exception ('52aa2007-c450-11e9-b5d4-a4d18cec433a'), another exception occurred:
>   File "/usr/local/lib/python3.10/dist-packages/django/core/handlers/exception.py", line 55, in inner
>     response = get_response(request)
>   File "/usr/local/lib/python3.10/dist-packages/django/core/handlers/base.py", line 197, in _get_response
>     response = wrapped_callback(request, *callback_args, **callback_kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/django/contrib/auth/decorators.py", line 23, in _wrapper_view
>     return view_func(request, *args, **kwargs)
>   File "/web_root/arches/arches/app/views/search.py", line 278, in export_results
>     export_files, export_info = exporter.export(format, report_link)
>   File "/web_root/arches/arches/app/search/search_export.py", line 160, in export
>     resource_obj = self.flatten_tiles(
>   File "/web_root/arches/arches/app/search/search_export.py", line 313, in flatten_tiles
>     node = self.get_node(nodeid)
>   File "/web_root/arches/arches/app/search/search_export.py", line 266, in get_node
>     self.node_lookup[nodeid] = models.Node.objects.get(pk=nodeid)
>   File "/usr/local/lib/python3.10/dist-packages/django/db/models/manager.py", line 87, in manager_method
>     return getattr(self.get_queryset(), name)(*args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/django/db/models/query.py", line 637, in get
>     raise self.model.DoesNotExist(
> 
> Exception Type: DoesNotExist at /search/export_results
> Exception Value: Node matching query does not exist.
> ```

I was able to adjust the search_export logic to not fail when a nodegroup isn't present. The output could be seen as confusing though; for example if the user exports a search result that contains 1 resource instance, and that resource instance is out-of-date, the export will succeed but the csv will be blank. 

I'll file a followup ticket to fix this. And I suppose this could spark a larger discussion of keeping cards/cards_x_nodes_x_widgets/edges/nodes/nodegroups between graph versions instead of deleting them 🤔  . It would result in more DB clutter but could be useful for cases such as this.

jacobtylerwalls commented on arches/app/models/models.py:
```diff
@@ -1152,6 +1197,10 @@ def __init__(self, *args, **kwargs):
             self.tileid = uuid.uuid4()
 
+    @property
+    def nodegroup(self):
```
Wondering if you considered making this a cached property (given how unlikely noedgroup-deletions-in-production are)?

> Be careful with your own custom properties - it is up to you to implement caching when required, for example using the [cached_property](https://docs.djangoproject.com/en/4.2/ref/utils/#django.utils.functional.cached_property) decorator.

https://docs.djangoproject.com/en/4.2/topics/db/optimization/#understand-cached-attributes

chrabyrd replied:
My gut says to shy away from caching in this instance because of the potential for unexpected behavior during graph-switching or updating

jacobtylerwalls replied:
sounds wise 👍

chrabyrd commented:
> Thanks for the walkthrough on the missing nodegroup flow, I get it now.
> 
> > As far as the nullguarding goes I'm inclined to just built an Exception into the Tile.nodegroup property instead of playing whack-a-mole with the callsites. WDYT?
> 
> I think I was wrongly assuming that some of these other workflows should still succeed if there are unmigrated historical tiles pointing to nonexistent nodes. But now that I see we want them to fail, then, yeah, let's not play whackamole. As you suggested something really simple like this would do:
> 
> ```diff
> diff --git a/arches/app/models/models.py b/arches/app/models/models.py
> index aa5a0d6e3..909e9ce5a 100644
> --- a/arches/app/models/models.py
> +++ b/arches/app/models/models.py
> @@ -1199,8 +1199,7 @@ class TileModel(models.Model):  # Tile
>  
>      @property
>      def nodegroup(self):
> -        nodegroup_query = NodeGroup.objects.filter(pk=self.nodegroup_id)
> -        return nodegroup_query.first()
> +        return NodeGroup.objects.get(pk=self.nodegroup_id)
>  
>      def is_fully_provisional(self):
>          return bool(self.provisionaledits and not any(self.data.values()))
> ```

Ah yeah after re-introducing the loud fail I remember why I took it out in the first place. It breaks serialization of Tiles and I'd like to not touch the long-untouched serializer if possible. I added it back in there with the most recent set of changes that un-breaks report for out of date Graphs.

apeters reviewed (changes requested):
`discard edits` never goes away after editing a card and then discarding the edits

<img width="493" alt="Screenshot 2024-04-12 at 9 55 53 AM" src="https://github.com/archesproject/arches/assets/823773/0789ea07-66f1-4719-9b1b-acc2405e21a7">

apeters commented:
We should add a cancel button to this popup.
<img width="607" alt="Screenshot 2024-04-12 at 12 17 14 PM" src="https://github.com/archesproject/arches/assets/823773/0ef55fa6-fb2d-4549-bada-5b3dcf83e16c">

chrabyrd commented:
> `discard edits` never goes away after editing a card and then discarding the edits
> 
> <img alt="Screenshot 2024-04-12 at 9 55 53 AM" width="493" src="https://private-user-images.githubusercontent.com/823773/322108761-0789ea07-66f1-4719-9b1b-acc2405e21a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2ODQ3OTksIm5iZiI6MTcxNDY4NDQ5OSwicGF0aCI6Ii84MjM3NzMvMzIyMTA4NzYxLTA3ODllYTA3LTY2ZjEtNDcxOS05YjFiLWFjYzI0MDVlMjFhNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwMlQyMTE0NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYWU4NDg5Mjg4NTU1NjAwNjA0MDk2NmZiMzAzMjkzZGVhMzc2ZGM5OGE4NmRmNTI4YTc2NzI3NTBkYWM4ZGQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.RWSbBWkH9YgaexYTklJzO4OlHNymJJqbHq9LOlRdBVQ">

This issue appears is dev/7.6.x -- and looks to be related to i18n. Agreed that it should be resolved but I believe it's out-of-scope for this PR. I ticketed the issue in #10873

chrabyrd commented on tests/fixtures/data/json/resource_export_business_data_truth.json:
```diff
@@ -41,6 +42,12 @@
             "data": {
               "677f236e-09cc-11e7-8ff7-6c4008b05c4c": null
             },
+            "nodegroup": {
+              "cardinality": "1",
+              "legacygroupid": null,
+              "nodegroupid": "677f236e-09cc-11e7-8ff7-6c4008b05c4c",
+              "parentnodegroup_id": null
+            },
```
@apeters 

This was fun, it took a while to track down. 

It's here: https://github.com/archesproject/arches/blob/dev/7.6.x/arches/app/utils/betterJSONSerializer.py#L209

Essentially because `nodegroup` is now a property on `TileModel` it gets serialized since `TileModel` doesn't override the default `serialize` method.

To fix this, I added `TileModel.serialize()` with a default exclude of `['nodegroup']`

jacobtylerwalls reviewed (approved):
Thanks for the updates, all my comments are addressed 👍

apeters reviewed (approved):
This looks great @chrabyrd !  thanks for sticking it out ❤️
